File: sdks/python/license_verifier.py
"""
Python Ed25519 License Verifier for Open-Core Dual-Licensing Engine
Requirements: cryptography >= 41.0.0
"""

import base64
import json
import datetime
from typing import Dict, Any, Optional
from cryptography.hazmat.primitives.asymmetric import ed25519
from cryptography.hazmat.primitives import serialization

class PythonLicenseVerifier:
    def __init__(self, public_key_pem: str):
        self.public_key = serialization.load_pem_public_key(public_key_pem.encode('utf-8'))

    def verify_license(self, license_key: str) -> Dict[str, Any]:
        try:
            parts = license_key.strip().split('.')
            if len(parts) != 2:
                return {"valid": False, "reason": "Invalid license key format."}

            payload_b64, signature_b64 = parts[0], parts[1]

            # Decode Base64URL
            payload_bytes = base64.urlsafe_b64decode(payload_b64 + '==')
            signature_bytes = base64.urlsafe_b64decode(signature_b64 + '==')

            # Verify Ed25519 Signature
            self.public_key.verify(signature_bytes, payload_b64.encode('utf-8'))

            # Parse Payload JSON
            payload = json.loads(payload_bytes.decode('utf-8'))

            expires_at = datetime.datetime.strptime(payload['expiresAt'], "%Y-%m-%d")
            now = datetime.datetime.utcnow()

            if now > expires_at:
                return {"valid": True, "active": False, "reason": f"License expired on {payload['expiresAt']}", "payload": payload}

            return {"valid": True, "active": True, "payload": payload}
        except Exception as e:
            return {"valid": False, "active": False, "reason": str(e)}

# Demo Usage
if __name__ == "__main__":
    print("Python Ed25519 License Verifier Module Loaded Successfully.")

File: sdks/go/license_verifier.go
package main

import (
	"crypto/ed25519"
	"encoding/base64"
	"encoding/pem"
	"encoding/json"
	"fmt"
	"strings"
	"time"
)

type CommercialLicensePayload struct {
	LicenseID     string `json:"licenseId"`
	CustomerName  string `json:"customerName"`
	CustomerEmail string `json:"customerEmail"`
	Tier          string `json:"tier"`
	MaxSeats      int    `json:"maxSeats"`
	MaxNodes      int    `json:"maxNodes"`
	IssuedAt      string `json:"issuedAt"`
	ExpiresAt     string `json:"expiresAt"`
}

type LicenseVerificationResult struct {
	Valid   bool                      `json:"valid"`
	Active  bool                      `json:"active"`
	Reason  string                    `json:"reason,omitempty"`
	Payload *CommercialLicensePayload `json:"payload,omitempty"`
}

func VerifyLicenseKey(licenseKey string, publicKeyPEM string) LicenseVerificationResult {
	block, _ := pem.Decode([]byte(publicKeyPEM))
	if block == nil {
		return LicenseVerificationResult{Valid: false, Reason: "Failed to decode PEM public key"}
	}

	pubKey := ed25519.PublicKey(block.Bytes)

	parts := strings.Split(strings.TrimSpace(licenseKey), ".")
	if len(parts) != 2 {
		return LicenseVerificationResult{Valid: false, Reason: "Invalid key format"}
	}

	payloadB64, sigB64 := parts[0], parts[1]

	sigBytes, err := base64.RawURLEncoding.DecodeString(sigB64)
	if err != nil {
		return LicenseVerificationResult{Valid: false, Reason: "Invalid signature encoding"}
	}

	isValid := ed25519.Verify(pubKey, []byte(payloadB64), sigBytes)
	if !isValid {
		return LicenseVerificationResult{Valid: false, Reason: "Signature verification failed"}
	}

	payloadBytes, err := base64.RawURLEncoding.DecodeString(payloadB64)
	if err != nil {
		return LicenseVerificationResult{Valid: false, Reason: "Invalid payload encoding"}
	}

	var payload CommercialLicensePayload
	if err := json.Unmarshal(payloadBytes, &payload); err != nil {
		return LicenseVerificationResult{Valid: false, Reason: "Failed to parse JSON payload"}
	}

	expirationDate, err := time.Parse("2006-01-02", payload.ExpiresAt)
	if err == nil && time.Now().After(expirationDate) {
		return LicenseVerificationResult{Valid: true, Active: false, Reason: "License expired", Payload: &payload}
	}

	return LicenseVerificationResult{Valid: true, Active: true, Payload: &payload}
}

func main() {
	fmt.Println("Go Ed25519 License Verifier SDK Module Loaded Successfully.")
}
